### PR TITLE
feat: Add pygame rendering capacity

### PR DESCRIPTION
**Pull request by josephmaa**
The goal of this PR is to add pygame rendering capacity to the gym environment. Future goals include working on adding side-scrolling capabilities for a better interactive viewing experience while building bridges. 

Currently, a known bug is that resizing the brick sizes is a bit off.

Screen Shot 2022-03-31 at 10 46 19 PM

### CODE REVIEW

File: gym_bridges/envs/bridges_env.py [hunk 1]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
...
+        if mode == "human":
+            mapping = {
+                BridgesEnv.StateType.GROUND: "@@",
+                BridgesEnv.StateType.BRICK: "[]",
+                BridgesEnv.StateType.EMPTY: "  ",
+            }
+            flat_repr = tuple([mapping[x] for x in self._state.flatten()])
+            print((("%s" * self.shape[1] + "\n") * self.shape[0]) % flat_repr)
+        elif mode == "pygame":
+            if not self._pygame_initialized:
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:328]: prefer to move this logic into a private function like 

def _initialize_pygame_if_necessary() -> None:
  if self._pygame_initialized:
     return
  // the rest of the code here to do the init work.

That makes this render() function shorter.

File: gym_bridges/envs/bridges_env.py [hunk 2]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
...
+                pygame.init()
+
+                self.screen = pygame.display.set_mode(
+                    (renderer_config.WINDOW_WIDTH, renderer_config.WINDOW_HEIGHT)
+                )
+                self.screen.fill(renderer_config.BLACK)
+
+                # Load the textures with a path relative to the game source code.
+                base_path = os.path.dirname(os.path.dirname(__file__))
+                textures_path = os.path.join(base_path, "renderer", "assets")
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:338]: I'm not fully sure what are the standard conventions for this.

I think I'd prefer a relative path that assumes the directory structure. Does it not work to do os.path.join("..", "renderer", "assets") ?

File: gym_bridges/envs/bridges_env.py [hunk 3]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
...
+
+                self.textures = {
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:340]: self._textures

File: gym_bridges/envs/bridges_env.py (hunk 1 above)
Comment (ldoshi) [reply gym_bridges/envs/bridges_env.py:328]: Per discussion, I think it's fine for the self._screen just to get initialized in the _if_necessary function.

File: gym_bridges/envs/bridges_env.py [hunk 4]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
...
+                        os.path.join(textures_path, "mossy_stone_bricks.png")
+                    ),
+                    BridgesEnv.StateType.EMPTY: pygame.image.load(
+                        os.path.join(textures_path, "light_blue_wool.png")
+                    ),
+                }
+
+                self._pygame_initialized = True
+
+            self.draw_state(screen=self.screen, textures=self.textures)
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:354]: self._draw_state( ) for private functions.

File: gym_bridges/envs/bridges_env.py (hunk 3 above)
Comment (ldoshi) [reply gym_bridges/envs/bridges_env.py:340]: base on the chat, I think _pygame_initialized should stay in init and self._screen and self._textures can live in the scope of this function.

File: gym_bridges/envs/bridges_env.py [hunk 5]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
...
+
+            # Drawing the state takes some time.
+            time.sleep(1)
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:359]: I guess the idea here is that if the user closes the window, they are done with their program run? I wonder if that's a bit aggressive vs just closing the window and staying closed.

File: gym_bridges/renderer/test_pygame_renderer.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+@pytest.mark.parametrize("width", [3, 9])
+def test_force_standard_config_render(width):
+    """The standard configuration only has ground at the bottom left
+    and right locations of the env."""
+    env = BridgesEnv(width=width, force_standard_config=True)
+    height = env.shape[0] - 1
+    initial_state = env.reset()
+    env.render(mode="pygame")
+    assert initial_state[height, 0] == BridgesEnv.StateType.GROUND
```
Comment (ldoshi) [inline gym_bridges/renderer/test_pygame_renderer.py:16]: I appreciate the desire to test very much!

I'm not sure this test does much though and I'm not sure how we'd test the pygame mechanically. I'd be ok with this being untested for now. 

One crazy idea could be if pygame lets one "save" the screen and then comparing to a golden file that it 'looks right'. I'm not convinced this effort (or approach) is necessary right now.

File: gym_bridges/envs/bridges_env.py (hunk 3 above)
Comment (ldoshi) [reply gym_bridges/envs/bridges_env.py:340]: (Full disclosure: My conventions weren't consistent when first writing this file so apologies for some inconsistency. Aiming for public attributes to not use _ and private ones to use _ )

File: gym_bridges/envs/bridges_env.py [hunk 7]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
 
         return self._state.copy()
 
+    def draw_state(
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:271]: def _draw_state()

File: gym_bridges/envs/bridges_env.py [hunk 8]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
...
+        self, screen: pygame.Surface, textures: dict[StateType, pygame.Surface]
+    ):
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:273]: return type -> None

File: gym_bridges/envs/bridges_env.py [hunk 9]
```diff
@@ -261,11 +268,96 @@ def reset(self, state=None, gap_count=None):
...
+            for y_coordinate, y_window_coordinate in zip(
+                y_coordinates, y_window_coordinates
+            ):
+                rect = pygame.Rect(
+                    x_window_coordinate,
+                    y_window_coordinate,
+                    block_size,
+                    block_size,
+                )
+                if (
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:302]: I think the following should work give or take:

for state_block, (x_window_coordinate, y_window_coordinate) in zip(self._state.flatten(), itertools.product(x_window_coordinates, y_window_coordinates)):
   rect = pygame.Rect(
                   x_window_coordinate,
                   y_window_coordinate,
                    block_size,
                    block_size,
               )
screen.blit(textures[state_block], rect)

File: gym_bridges/envs/bridges_env.py (hunk 2 above)
Comment (josephmaa) [reply gym_bridges/envs/bridges_env.py:323]: From looking at other stack overflow posts, I don't think there's a better way to access the relative parent directory from python. The `Pathlib` module essentially does the same thing, but we've been using `os` in the rest of the project so far.

File: gym_bridges/envs/bridges_env.py (hunk 5 above)
Comment (josephmaa) [reply gym_bridges/envs/bridges_env.py:344]: I was following[ the convention here](https://stackoverflow.com/questions/19882415/closing-pygame-window).

File: gym_bridges/envs/test_bridges_env.py [hunk 10]
```diff
@@ -9,6 +9,16 @@
 from gym_bridges.envs.bridges_env import BridgesEnv
 
 
+@pytest.mark.parametrize("width", [3, 9])
+def test_force_standard_config_render(width):
+    """The standard configuration only has ground at the bottom left
+    and right locations of the env."""
+    env = BridgesEnv(width=width, force_standard_config=True)
+    env.reset()
```
Comment (josephmaa) [inline gym_bridges/envs/test_bridges_env.py:17]: This test doesn't run correctly without `env.reset()`. Is it expected that every `BridgesEnv` instance will be reset before being renderer?

File: gym_bridges/envs/bridges_env.py (hunk 5 above)
Comment (ldoshi) [reply gym_bridges/envs/bridges_env.py:341]: SGTM

File: gym_bridges/envs/test_bridges_env.py (hunk 10 above)
Comment (ldoshi) [reply gym_bridges/envs/test_bridges_env.py:17]: I think so... reset() before doing anything else seems to be the convention per the small code snippet here and a few other places I've seen.

https://gym.openai.com/

File: gym_bridges/envs/bridges_env.py [hunk 11]
```diff
@@ -261,11 +268,79 @@ def reset(self, state=None, gap_count=None):
...
+            rect = pygame.Rect(
+                x_window_coordinate,
+                y_window_coordinate,
+                block_size,
+                block_size,
+            )
+            self._screen.blit(self._textures[state_block], rect)
+
+    def _initialize_pygame_if_necessary(self) -> None:
+        if self._initialize_pygame:
```
Comment (ldoshi) [inline gym_bridges/envs/bridges_env.py:298]: optional (but preferred) nit: I  like the style of 

if not self._initialize_pygame:
  return

pygame.init() etc


benefits:
If it's self._initialize_pygame is false, it's very clear that we need not read further.

This also means the entire function isn't a nested if-statement.

File: gym_bridges/envs/test_bridges_env.py [hunk 12]
```diff
@@ -6,7 +6,17 @@
...
+
+
+@pytest.mark.parametrize("width", [3, 9])
+def test_force_standard_config_render(width):
+    """The standard configuration only has ground at the bottom left
+    and right locations of the env."""
+    env = BridgesEnv(width=width, force_standard_config=True)
+    env.reset()
+    env.render(mode="pygame")
+    time.sleep(5)
```
Comment (ldoshi) [inline gym_bridges/envs/test_bridges_env.py:19]: Prefer delete the sleep. I realize this is nice as a human to see it working. But now that's it's working, the main purpose of the test is that the code is not broken. If running in CI (in practice or in theory), no one is looking at the output.

**Pull request merged**